providers/gitlab.go

```diff
@@ -206,15 +289,70 @@ func (p *GitLabProvider) EnrichSessionState(ctx context.Context, s *sessions.Ses
 		return fmt.Errorf("user email is not verified")
 	}
 
-	// Check group membership
-	// TODO (@NickMeves) - Refactor to Authorize
-	err = p.verifyGroupMembership(userInfo)
-	if err != nil {
-		return fmt.Errorf("group membership check failed: %v", err)
-	}
+	p.addGroupsToSession(ctx, s)
+
+	p.addProjectsToSession(ctx, s)
 
 	s.User = userInfo.Username
 	s.Email = userInfo.Email
 
 	return nil
+
+}
+
+// addGroupsToSession projects into session.Groups
+func (p *GitLabProvider) addGroupsToSession(ctx context.Context, s *sessions.SessionState) {
+	// Iterate over projects, check if oauth2-proxy can get project information on behalf of the user
+	for _, group := range p.Groups {
+		s.Groups = append(s.Groups, fmt.Sprintf("group:%s", group))
+	}
+
+}
+
+// addProjectsToSession projects into session.Groups
+func (p *GitLabProvider) addProjectsToSession(ctx context.Context, s *sessions.SessionState) {
+	// Iterate over projects, check if oauth2-proxy can get project information on behalf of the user
+	for _, project := range p.Projects {
+		projectInfo, err := p.getProjectInfo(ctx, s, project.Name)
+
+		if err == nil && !projectInfo.Archived {
+			// try first with project access
+			perms := projectInfo.Permissions.ProjectAccess
+			if perms == nil {
+				// use group project access as fallback
+				perms = projectInfo.Permissions.GroupAccess
+			}
+
+			if perms.AccessLevel >= project.AccessLevel {
+				s.Groups = append(s.Groups, fmt.Sprintf("project:%s", project.Name))
+			} else {
+				logger.Errorf("Warning: user %s do not have the minimum required access level %s", s.Email, project.Name)
```

Comment:
Whoops, missed the grammar error in my first recommended change:
```suggestion
				logger.Errorf("Warning: user %s does not have the minimum required access level %s", s.Email, project.Name)
```

Reply:
Might want to make it more explicit that this is the access level for a project, also, could use `%q` to quote the strings being entered as well, WDYT?
```suggestion
				logger.Errorf("Warning: user %q does not have the minimum required access level for project %q", s.Email, project.Name)
```